### PR TITLE
MRG: Fix metadata

### LINKED ISSUE #4820: BUG: metadata and droplog
`metadata` does not take dropped epochs into account

e.g.

```python
import numpy as np
import pandas as pd
# make up raw
raw_data = np.random.randn(10, 1000)
info = mne.create_info(10, 1000.)
raw = mne.io.RawArray(raw_data, info)
# make up events with metadata
events = [[0, 0, 1], [100, 0, 1], [200, 0, 1], [300, 0, 1]]
metadata = pd.DataFrame([dict(idx=idx) for idx in range(len(events))])
epochs = mne.Epochs(raw, events=np.array(events), tmin=-.050, tmax=.100,
                    metadata=metadata)
print('bad epochs:', np.where(epochs.drop_log))
epochs.drop_bad()
print('n_epochs:%i, n_metadata:%i' % (len(epochs), len(epochs.metadata)))
print(epochs.metadata)
```

cc @jona-sassenhagen @choldgraf

### LINKED ISSUE #4821: BUG: metadata update field
The user cannot update the metadata field:

```python

import numpy as np
import pandas as pd
epo = mne.Epochs(mne.io.RawArray(np.random.randn(10, 1000),
                                 mne.create_info(10, 1000.)),
                 events=np.array([[100, 0, 1], [200, 0, 1], [300, 0, 1]]),
                 tmin=-.050, tmax=.100,
                 metadata=pd.DataFrame([dict(original_key=idx) for idx in range(3)]))
epo.metadata['new_key'] = 1
print(epo['original_key'])  # works
print(epo.metadata['new_key'])  # does not work
print(epo['new_key'])  # doesnt work
```

cc @choldgraf @jona-sassenhagen

### LINKED ISSUE #4823: FIX: metadata index behavior does not comply with pandas
the raw index is reset by default. This makes the user lose track of the original indices of the metadata. #bad

```python
import mne
import numpy as np
import pandas as pd
epo = mne.Epochs(mne.io.RawArray(np.random.randn(10, 1000),
                                 mne.create_info(10, 1000.)),
                 events=np.array([[100, 0, 1], [200, 0, 1], [300, 0, 1]]),
                 tmin=-.050, tmax=.100,
                 metadata=pd.DataFrame([dict(original_key=idx) for idx in range(3)]))

epo[1:].metadata  # the raw index is reset by default. 
```

### PR DESCRIPTION
**Pull request by larsoner**
Closes #4820.
Closes #4821.
Closes #4823.

@kingjr can you confirm everything works as you expect now?

### CODE REVIEW

**Comment by kingjr** [comment]
Perfect, thanks @larsoner !

File: examples/preprocessing/plot_metadata_query.py [hunk 1]
```diff
@@ -63,8 +63,11 @@ def plot_query_results(query):
 # First we'll create our metadata object. This should be a
 # :class:`pandas.DataFrame` with each row corresponding to an event.
 #
-# .. warning:: The Dataframe Index can change during MNE I/O operations, so
-#              do not rely on it to query your metadata.
+# .. warning:: The Dataframe index will be controlled by MNE to track
+#              ``epochs.selection``, so do not manually set this value.
```
Comment (kingjr) [inline examples/preprocessing/plot_metadata_query.py]: ambiguous wording: which value? `index` or `epochs.selection`

**Comment by larsoner** [comment]
There are I/O tests already on ~L2282 (and they happen after I do the dropping on ~L2272) so I'm not sure if you want something more or just didn't realize they are there already...

**Comment by kingjr** [comment]
Considering the importance of #4820, shall we backport this fix to 0.15?

**Comment by larsoner** [comment]
This wasn't in 0.15, we purposely didn't merge until releasing 0.15 to catch things like this

**Comment by kingjr** [comment]
Good call!

**Comment by larsoner** [comment]
@kingjr verbosity and wording in example fixed, feel free to look / try again

**Comment by kingjr** [comment]
LGTM

**Comment by agramfort** [comment]
@jona-sassenhagen wanna have a look?

+1 for MRG on my side

**Comment by jona-sassenhagen** [comment]
Can't find any errors. So Appveyor is still stuck ..?

Guess I'll merge anyways.

**Comment by jona-sassenhagen** [comment]
Thanks Eric!

**Pull request merged**